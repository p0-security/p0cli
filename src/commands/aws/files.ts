/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { AwsCredentials } from "../../plugins/aws/types";
import * as ini from "ini";
import * as fs from "node:fs/promises";
import * as os from "node:os";
import path from "node:path";
import tmp from "tmp-promise";

const AWS_CONFIG_PATH = path.join(os.homedir(), ".aws");
export const AWS_CONFIG_FILE = path.join(AWS_CONFIG_PATH, "config");
export const AWS_CREDENTIALS_FILE = path.join(AWS_CONFIG_PATH, "credentials");

// Reference documentation: https://docs.aws.amazon.com/sdkref/latest/guide/file-format.html

/**
 * Reads in an AWS CLI configuration file, which is formatted as INI text, and
 * returns an arbitrary object representing the contents.
 *
 * @param path Path of the file to read
 * @returns Arbitrary object representing the contents of the file, or an empty
 * object if the file is empty or does not exist
 */
export const readIniFile = async (
  path: string
): Promise<{ [key: string]: any }> => {
  try {
    const data = await fs.readFile(path, { encoding: "utf-8" });
    return data ? ini.parse(data) : {};
  } catch (err: any) {
    if (err.code === "ENOENT") {
      return {};
    }

    throw err;
  }
};

/**
 * This function writes an arbitrary object as INI-formatted text to a file
 * atomically by first writing the data to a temporary file then moving the
 * temporary file on top of the target file. This minimizes the chance that an
 * exception, system crash, or other similar event will leave the file in a
 * corrupted state; this is important since we're mucking around with the AWS
 * CLI's configuration files.
 *
 * @param path Path of the (permanent) file to write to
 * @param obj Arbitrary object to convert to INI-formatted text and write to the
 * file
 * @param iniEncodeOptions Options to pass to the INI encoding library
 */
export const atomicWriteIniFile = async (
  path: string,
  obj: any,
  iniEncodeOptions?: ini.EncodeOptions
): Promise<void> => {
  const data = ini.stringify(obj, iniEncodeOptions);

  // Permissions will be moved along with the file
  const { path: tmpPath } = await tmp.file({ mode: 0o600, prefix: "p0cli-" });

  await fs.writeFile(tmpPath, data, { encoding: "utf-8" });
  await fs.rename(tmpPath, path);
};

export const writeAwsTempCredentials = async (
  profileName: string,
  awsCredentials: AwsCredentials
) => {
  const credentials = await readIniFile(AWS_CREDENTIALS_FILE);

  credentials[profileName] = {
    aws_access_key_id: awsCredentials.AWS_ACCESS_KEY_ID,
    aws_secret_access_key: awsCredentials.AWS_SECRET_ACCESS_KEY,
    aws_session_token: awsCredentials.AWS_SESSION_TOKEN,
  };

  // The credentials file is formatted with whitespace before and after the `=`
  await atomicWriteIniFile(AWS_CREDENTIALS_FILE, credentials, {
    whitespace: true,
  });
};

export const writeAwsConfigProfile = async (
  profileName: string,
  profileConfig: any
) => {
  const config = await readIniFile(AWS_CONFIG_FILE);

  config[`profile ${profileName}`] = profileConfig;

  await atomicWriteIniFile(AWS_CONFIG_FILE, config);
};
