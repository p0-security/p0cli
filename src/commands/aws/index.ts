/** Copyright Â© 2024-present P0 Security 

This file is part of @p0security/p0cli

@p0security/p0cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/p0cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/p0cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { role } from "./role";
import yargs from "yargs";

const awsCommands = [role];

const awsArgs = (yargs: yargs.Argv) => {
  const base = yargs
    .option("account", {
      type: "string",
      describe: "AWS account ID or alias (or set P0_AWS_ACCOUNT)",
    })
    .env("P0_AWS");
  return awsCommands.reduce((m, c) => c(m), base).demandCommand(1);
};

export const awsCommand = (yargs: yargs.Argv) =>
  yargs.command("aws", "Execute AWS commands", awsArgs);
