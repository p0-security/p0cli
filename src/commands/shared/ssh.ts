/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { decodeProvisionStatus } from ".";
import { createKeyPair } from "../../common/keys";
import { fetchIntegrationConfig } from "../../drivers/api";
import { getContactMessage } from "../../drivers/config";
import { print2 } from "../../drivers/stdio";
import { awsSshProvider } from "../../plugins/aws/ssh";
import { azureSshProvider } from "../../plugins/azure/ssh";
import { gcpSshProvider } from "../../plugins/google/ssh";
import { selfHostedSshProvider } from "../../plugins/self-hosted/ssh";
import { SshConfig } from "../../plugins/ssh/types";
import { Authn } from "../../types/identity";
import { PermissionRequest } from "../../types/request";
import {
  CliSshRequest,
  PluginSshRequest,
  SshProvider,
  SupportedSshProvider,
  SupportedSshProviders,
} from "../../types/ssh";
import { request } from "./request";
import { pick } from "lodash";
import { sys } from "typescript";
import yargs from "yargs";

export type BaseSshCommandArgs = {
  sudo?: boolean;
  reason?: string;
  parent?: string;
  provider?: SupportedSshProvider;
  debug?: boolean;
  sshOptions?: string[];
};

export type ScpCommandArgs = BaseSshCommandArgs & {
  source: string;
  destination: string;
};

export type SshCommandArgs = BaseSshCommandArgs & {
  sudo?: boolean;
  destination: string;
  arguments: string[];
  command?: string;
};

export type SshResolveCommandArgs = SshCommandArgs & {
  quiet?: boolean;
};

export type SshProxyCommandArgs = {
  destination: string;
  port: string;
  provider: "aws" | "azure" | "gcloud";
  requestJson: string;
  debug?: boolean;
  identityFile: string;
};

export type CommandArgs = ScpCommandArgs | SshCommandArgs;

export type SshRequestOptions = {
  approvedOnly?: boolean;
  quiet?: boolean;
};

export type SshAdditionalSetup = {
  /** A list of SSH configuration options, as would be used after '-o' in an SSH command */
  sshOptions: string[];

  /** The path to the private key file to use for the SSH connection, instead of the default P0 CLI managed key */
  identityFile?: string;

  /** The port to connect to, overriding the default */
  port?: string;

  /** Perform any teardown required after the SSH command exits but before terminating the P0 CLI */
  teardown: () => Promise<void>;
};

export const SSH_PROVIDERS: Record<
  SupportedSshProvider,
  SshProvider<any, any, any, any>
> = {
  aws: awsSshProvider,
  azure: azureSshProvider,
  gcloud: gcpSshProvider,
  "self-hosted": selfHostedSshProvider,
};

const validateSshInstall = async (
  authn: Authn,
  args: yargs.ArgumentsCamelCase<BaseSshCommandArgs>
) => {
  const configDoc = await fetchIntegrationConfig<{ config: SshConfig }>(
    authn,
    "ssh",
    args.debug
  );
  const configItems = configDoc?.config["iam-write"];

  const providersToCheck = args.provider
    ? [args.provider]
    : SupportedSshProviders;

  const items = Object.entries(configItems ?? {}).filter(
    ([key, value]) =>
      value.state == "installed" &&
      providersToCheck.some((prefix) => key.startsWith(prefix))
  );

  if (items.length === 0) {
    throw "This organization is not configured for SSH access";
  }
};

export const isSudoCommand = (args: { sudo?: boolean; command?: string }) =>
  args.sudo || args.command === "sudo";

export const provisionRequest = async (
  authn: Authn,
  args: yargs.ArgumentsCamelCase<BaseSshCommandArgs>,
  destination: string,
  options?: SshRequestOptions
) => {
  await validateSshInstall(authn, args);

  const { publicKey, privateKey } = await createKeyPair();

  const makeRequest = async (requestOptions?: { forceSudo: boolean }) => {
    return await request("request")<PermissionRequest<PluginSshRequest>>(
      {
        ...pick(args, "$0", "_"),
        arguments: [
          "ssh",
          "session",
          destination,
          "--public-key",
          publicKey,
          ...(options?.approvedOnly ? ["--approved"] : []),
          ...(args.provider ? ["--provider", args.provider] : []),
          ...(requestOptions?.forceSudo || isSudoCommand(args)
            ? ["--sudo"]
            : []),
          ...(args.reason ? ["--reason", args.reason] : []),
          ...(args.parent ? ["--parent", args.parent] : []),
        ],
        wait: true,
        debug: args.debug,
      },
      authn,
      { message: options?.quiet ? "quiet" : "approval-required" }
    );
  };

  const requestErrorHandler = (err: any) => {
    if (typeof err === "string") {
      print2(err);
      if (
        err.startsWith("Could not find any instances matching") &&
        err.includes("@")
      ) {
        print2(
          "Hint: The instance-name appears to contain a username AND a hostname; only the hostname is required."
        );
      }
    }
    sys.exit(1);
  };

  let response;
  if (options?.approvedOnly) {
    // Try first with sudo
    try {
      response = await makeRequest({ forceSudo: true }).catch(
        requestErrorHandler
      );
    } catch (error) {
      // If that fails, try without sudo
      if (args.debug) {
        print2("Request with sudo failed, retrying without sudo");
      }
      response = await makeRequest().catch(requestErrorHandler);
    }
  } else {
    // Normal behavior when not approvedOnly
    response = await makeRequest().catch(requestErrorHandler);
  }

  if (!response) {
    if (!options?.quiet) {
      print2("Did not receive access ID from server");
    }
    return;
  }

  const { id, isPreexisting } = response;

  const message = isPreexisting
    ? "Existing access found.  Connecting to instance."
    : "Waiting for access to be provisioned";
  print2(message);

  const result = await decodeProvisionStatus<PluginSshRequest>(
    response.request
  );

  if (!result) sys.exit(1);

  return {
    requestId: id,
    provisionedRequest: response.request,
    publicKey,
    privateKey,
  };
};

const pluginToCliRequest = async (
  request: PermissionRequest<PluginSshRequest>,
  options: { debug?: boolean; publicKey: string }
): Promise<PermissionRequest<CliSshRequest>> =>
  await SSH_PROVIDERS[request.permission.provider].toCliRequest(
    request as any,
    options
  );

export const prepareRequest = async (
  authn: Authn,
  args: yargs.ArgumentsCamelCase<BaseSshCommandArgs>,
  destination: string,
  options?: SshRequestOptions
) => {
  const result = await provisionRequest(authn, args, destination, options);
  if (!result) {
    throw `Server did not return a request id. ${getContactMessage()}`;
  }

  const { requestId, publicKey, provisionedRequest } = result;

  const sshProvider = SSH_PROVIDERS[provisionedRequest.permission.provider];

  await sshProvider.submitPublicKey?.(
    authn,
    provisionedRequest,
    requestId,
    publicKey,
    args.debug
  );

  await sshProvider.ensureInstall();

  const cliRequest = await pluginToCliRequest(provisionedRequest, {
    ...args,
    publicKey,
  });

  const request = sshProvider.requestToSsh(cliRequest);

  const sshHostKeys = await sshProvider.saveHostKeys?.(request, args);

  return { ...result, request, sshProvider, provisionedRequest, sshHostKeys };
};
