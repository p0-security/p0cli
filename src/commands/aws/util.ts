/** Copyright Â© 2025-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { print2 } from "../../drivers/stdio";
import { Authn } from "../../types/identity";
import { PluginRequest } from "../../types/request";
import { waitForProvisioning } from "../shared";
import { request } from "../shared/request";
import yargs from "yargs";

export const ACCESS_EXISTS_ERROR_MESSAGE =
  "This principal already has this access";

export const provisionRequest = async (
  argv: yargs.ArgumentsCamelCase<{
    arguments: string[];
    wait?: boolean;
  }>,
  authn: Authn
) => {
  const response = await request("request")(argv, authn, {
    message: "approval-required",
  });

  if (!response) {
    print2("Did not receive access ID from server");
    return;
  }

  const { id, isPreexisting } = response;

  print2(
    !isPreexisting
      ? "Waiting for access to be provisioned"
      : "Existing access found. Connecting to instance."
  );

  const provisionedRequest = await waitForProvisioning<PluginRequest>(
    authn,
    id
  );

  return { requestId: id, provisionedRequest };
};
