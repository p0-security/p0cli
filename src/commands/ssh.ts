/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { authenticate } from "../drivers/auth";
import { guard } from "../drivers/firestore";
import { sshOrScp } from "../plugins/ssh";
import { SshCommandArgs, prepareRequest } from "./shared/ssh";
import yargs from "yargs";

export const sshCommand = (yargs: yargs.Argv) =>
  yargs.command<SshCommandArgs>(
    "ssh <destination> [command [arguments..]]",
    "SSH into a virtual machine",
    (yargs) =>
      yargs
        .positional("destination", {
          type: "string",
          demandOption: true,
        })
        .positional("command", {
          type: "string",
          describe: "Pass command to the shell",
        })
        .positional("arguments", {
          describe: "Command arguments",
          array: true,
          string: true,
          default: [] as string[],
        })
        .option("sudo", {
          type: "boolean",
          describe: "Add user to sudoers file",
        })
        .option("L", {
          type: "string",
          // Copied from `man ssh`
          describe:
            "Specifies that connections to the given TCP port or Unix socket on the local (client) host are to be forwarded to the given host and port, or Unix socket, on the remote side.",
        })
        .option("R", {
          type: "string",
          // Copied from `man ssh`
          describe:
            "Specifies that connections to the given TCP port or Unix socket on the remote (server) host are to be forwarded to the local side.",
        })
        .option("N", {
          type: "boolean",
          describe:
            "Do not execute a remote command. Useful for forwarding ports.",
        })
        .option("A", {
          type: "boolean",
          describe:
            "Enables forwarding of connections from an authentication agent such as ssh-agent",
        })
        .option("o", {
          type: "string",
          describe:
            "Can be used to give options in the format used in the SSH configuration file.",
        })
        // Match `p0 request --reason`
        .option("reason", {
          describe: "Reason access is needed",
          type: "string",
        })
        .option("account", {
          type: "string",
          describe: "The account on which the instance is located",
        })
        .option("provider", {
          type: "string",
          describe: "The cloud provider where the instance is hosted",
          choices: ["aws", "gcloud"],
        })
        .option("debug", {
          type: "boolean",
          describe: "Print debug information.",
        }),
    guard(sshAction)
  );

/** Connect to an SSH backend
 *
 * Implicitly gains access to the SSH resource if required.
 *
 * Supported SSH mechanisms:
 * - AWS EC2 via SSM with Okta SAML
 */
const sshAction = async (args: yargs.ArgumentsCamelCase<SshCommandArgs>) => {
  // Prefix is required because the backend uses it to determine that this is an AWS request
  const authn = await authenticate();

  const { request, privateKey, sshProvider } = await prepareRequest(
    authn,
    args,
    args.destination
  );

  await sshOrScp({
    authn,
    request,
    cmdArgs: args,
    privateKey,
    sshProvider,
  });
};
