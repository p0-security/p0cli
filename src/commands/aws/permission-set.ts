/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { fsShutdownGuard } from "../../drivers/firestore";
import { getAwsConfig } from "../../plugins/aws/config";
import { assumeRoleWithIdc } from "../../plugins/aws/idc";
import { Authn } from "../../types/identity";
import { provisionRequest } from "../shared/request";
import { AssumeCommandArgs, AssumePermissionSetCommandArgs } from "./types";
import { printAwsCredentials } from "./util";
import { pick } from "lodash";
import yargs from "yargs";

export const permissionSet = (
  yargs: yargs.Argv<AssumeCommandArgs>,
  authn: Authn
) =>
  yargs.command(
    "permission-set",
    "Interact with AWS permission sets",
    (yargs) =>
      yargs.command(
        "assume <permission-set>",
        "Assume an AWS permission set",
        (y: yargs.Argv<AssumeCommandArgs>) =>
          y.positional("permission-set", {
            type: "string",
            demandOption: true,
            describe: "An AWS permission set name",
          }),
        fsShutdownGuard((argv) => oktaAwsAssumePermissionSet(argv, authn))
      )
  );

const oktaAwsAssumePermissionSet = async (
  argv: yargs.ArgumentsCamelCase<AssumePermissionSetCommandArgs>,
  authn: Authn
) => {
  const { account, permissionSet } = argv;
  const { config } = await getAwsConfig(authn, account);

  if (config.login?.type !== "idc") {
    throw new Error(
      `Unexpected login type. Expected IDC to be enabled for account ${account}`
    );
  }

  const { login } = config;

  const requestCommand = buildPermissionSetRequestCommand(argv);

  await provisionRequest(requestCommand, authn);

  const awsCredential = await assumeRoleWithIdc({
    accountId: config.id,
    permissionSet,
    idc: { id: login.identityStoreId, region: login.idcRegion },
  });

  const command = `p0 aws${argv.account ? ` --account ${argv.account}` : ""} permission-set assume ${argv.permissionSet}`;
  printAwsCredentials(awsCredential, command);
};

const buildPermissionSetRequestCommand = (
  argv: yargs.ArgumentsCamelCase<AssumePermissionSetCommandArgs>
): yargs.ArgumentsCamelCase<{
  arguments: string[];
  wait?: boolean;
}> => {
  return {
    ...pick(argv, "$0", "_"),
    arguments: [
      "aws",
      "permission-set",
      argv.permissionSet,
      ...(argv.reason ? ["--reason", argv.reason] : []),
      ...(argv.account ? ["--account", argv.account] : []),
    ],
    wait: true,
  };
};
