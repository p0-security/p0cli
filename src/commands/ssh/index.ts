import { authenticate } from "../../drivers/auth";
import { doc, guard } from "../../drivers/firestore";
import { AWS_API_VERSION } from "../../plugins/aws/api";
import { AwsCredentials } from "../../plugins/aws/types";
import { assumeRoleWithOktaSaml } from "../../plugins/okta/aws";
import { Authn } from "../../types/identity";
import {
  DENIED_STATUSES,
  DONE_STATUSES,
  ERROR_STATUSES,
  Request,
} from "../../types/request";
import { sleep } from "../../util";
import { request } from "../request";
import {
  SSMClient,
  StartSessionCommand,
  StartSessionCommandInput,
} from "@aws-sdk/client-ssm";
import { onSnapshot } from "firebase/firestore";
import { pick } from "lodash";
import { spawn } from "node:child_process";
import yargs from "yargs";

const awsRequestPrefix = "AWS:";

/** Maximum amount of time to wait after access is approved to wait for access
 *  to be configured
 */
const GRANT_TIMEOUT_MILLIS = 60e3;

/** Maximum amount of time to wait after access is configured in AWS to wait
 *  for access to propagate through AWS
 */
const ACCESS_TIMEOUT_MILLIS = 30e3;
/** Polling interval for the above check */
const ACCESS_CHECK_INTERVAL_MILLIS = 200;

const INSTANCE_ARN_PATTERN =
  /^arn:aws:ssm:([^:]+):([^:]+):managed-instance\/([^:]+)$/;

const ssmDocumentArn = (requestId: string) =>
  `P0SshAsUser-${requestId}-nathan_brahms`;

type AwsSsh = {
  spec: {
    arn: string;
  };
  type: "session";
};

type SsmArgs = {
  instance: string;
  region: string;
  requestId: string;
  documentName: string;
  credential: AwsCredentials;
};

export const sshCommand = (yargs: yargs.Argv) =>
  yargs.command<{ instance: string }>(
    "ssh <instance>",
    "SSH into a virtual machine",
    (yargs) =>
      yargs.positional("instance", {
        type: "string",
        demandOption: true,
      }),
    guard(ssh)
  );

const waitForProvisioning = async (authn: Authn, requestId: string) => {
  let cancel: NodeJS.Timeout | undefined = undefined;
  const result = await new Promise<Request<AwsSsh>>((resolve, reject) => {
    let isResolved = false;
    const unsubscribe = onSnapshot<Request<AwsSsh>, object>(
      doc(`o/${authn.identity.org.tenantId}/permission-requests/${requestId}`),
      (snap) => {
        const data = snap.data();
        if (!data) return;
        if (DONE_STATUSES.includes(data.status as any)) {
          resolve(data);
        } else if (DENIED_STATUSES.includes(data.status as any)) {
          reject("Your access request was denied");
        } else if (ERROR_STATUSES.includes(data.status as any)) {
          reject(
            "Your access request encountered an error (see Slack for details)"
          );
        } else {
          return;
        }
        isResolved = true;
        unsubscribe();
      }
    );
    cancel = setTimeout(() => {
      if (!isResolved) {
        unsubscribe();
        reject("Timeout awaiting SSH access grant");
      }
    }, GRANT_TIMEOUT_MILLIS);
  });
  clearTimeout(cancel);
  return result;
};

const waitForAccess = async (args: SsmArgs) => {
  const start = Date.now();
  let lastError: any = undefined;
  const client = new SSMClient({
    apiVersion: AWS_API_VERSION,
    region: args.region,
    credentials: {
      accessKeyId: args.credential.AWS_ACCESS_KEY_ID,
      secretAccessKey: args.credential.AWS_SECRET_ACCESS_KEY,
      sessionToken: args.credential.AWS_SESSION_TOKEN,
    },
  });
  const commandInput: StartSessionCommandInput = {
    Target: args.instance,
    DocumentName: args.documentName,
  };
  while (Date.now() - start < ACCESS_TIMEOUT_MILLIS) {
    try {
      // We don't use this response for anything; we just need this to succeed
      await client.send(
        new StartSessionCommand({
          ...commandInput,
          Reason: "Test connectivity",
        })
      );
      return commandInput;
    } catch (error: any) {
      if (error.__type === "AccessDeniedException") {
        lastError = error;
        await sleep(ACCESS_CHECK_INTERVAL_MILLIS);
      } else throw error;
    }
  }
  throw lastError;
};

const spawnSsmNode = async (
  args: Pick<SsmArgs, "region" | "credential">,
  commandInput: StartSessionCommandInput
) =>
  new Promise((resolve, reject) => {
    if (!commandInput.Target || !commandInput.DocumentName) {
      reject("Command input is missing required fields: Target, DocumentName");
      return;
    }
    const child = spawn(
      "/usr/bin/env",
      [
        "aws",
        "ssm",
        "start-session",
        "--target",
        commandInput.Target,
        "--document-name",
        commandInput.DocumentName,
      ],
      {
        env: {
          ...process.env,
          ...args.credential,
          AWS_DEFAULT_REGION: args.region,
        },
        stdio: "inherit",
      }
    );

    const exitListener = child.on("exit", (code) => {
      exitListener.unref();
      console.error("SSH session terminated");
      resolve(code);
    });
  });

const ssm = async (authn: Authn, request: Request<AwsSsh> & { id: string }) => {
  const match = request.permission.spec.arn.match(INSTANCE_ARN_PATTERN);
  if (!match) throw "Did not receive a properly formatted instance identifier";
  const [, region, account, instance] = match;

  const credential = await assumeRoleWithOktaSaml(authn, {
    account,
    role: request.generatedRoles[0]!.role,
  });
  const args = {
    instance: instance!,
    region: region!,
    documentName: request.generated.documentName,
    requestId: request.id,
    credential,
  };
  const commandInput = await waitForAccess(args);
  await spawnSsmNode(args, commandInput);
};

const ssh = async (args: yargs.ArgumentsCamelCase<{ instance: string }>) => {
  // Prefix is required because the backend uses it to determine that this is an AWS request
  const authn = await authenticate();
  const response = await request(
    {
      ...pick(args, "$0", "_"),
      arguments: ["ssh", "session", args.instance, "--provider", "aws"],
      wait: true,
    },
    authn,
    { message: "approval-required" }
  );
  if (!response) {
    console.error("Did not receive access ID from server");
    return;
  }
  const { id, isPreexisting } = response;
  if (!isPreexisting) console.error("Waiting for access to be provisioned");
  const requestData = await waitForProvisioning(authn, id);
  await ssm(authn, { ...requestData, id });
};
