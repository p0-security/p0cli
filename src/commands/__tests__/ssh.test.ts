/** Copyright Â© 2024-present P0 Security

This file is part of @p0security/cli

@p0security/cli is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 of the License.

@p0security/cli is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with @p0security/cli. If not, see <https://www.gnu.org/licenses/>.
**/
import { TEST_PUBLIC_KEY } from "../../common/__mocks__/keys";
import { fetchCommand } from "../../drivers/api";
import { print1, print2 } from "../../drivers/stdio";
import { AwsSshGenerated, AwsSshPermission } from "../../plugins/aws/types";
import { sshOrScp } from "../../plugins/ssh";
import { mockGetDoc } from "../../testing/firestore";
import { sleep } from "../../util";
import { sshCommand } from "../ssh";
import { onSnapshot } from "firebase/firestore";
import { noop, omit } from "lodash";
import yargs from "yargs";

jest.mock("../../drivers/api");
jest.mock("../../drivers/auth");
jest.mock("../../drivers/stdio");
jest.mock("../../plugins/ssh");
jest.mock("../../common/keys");

const mockFetchCommand = fetchCommand as jest.Mock;
const mockSshOrScp = sshOrScp as jest.Mock;
const mockPrint1 = print1 as jest.Mock;
const mockPrint2 = print2 as jest.Mock;

const MOCK_PERMISSION: AwsSshPermission = {
  provider: "aws",
  publicKey: TEST_PUBLIC_KEY,
  region: "region",
  alias: "alias",
  resource: {
    account: "accountId",
    accountId: "accountId",
    arn: "arn",
    idcRegion: "idcRegion",
    idcId: "idcId",
    name: "name",
    userName: "userName",
    instanceId: "instanceId",
  },
};

const MOCK_GENERATED = {
  resource: {
    name: "name",
  },
  linuxUserName: "linuxUserName",
} as AwsSshGenerated;

const MOCK_REQUEST = {
  status: "DONE",
  generated: MOCK_GENERATED,
  permission: MOCK_PERMISSION,
};

mockGetDoc({
  "iam-write": {
    ["aws:test-account"]: {
      state: "installed",
    },
  },
});

describe("ssh", () => {
  describe.each([
    ["persistent", true],
    ["ephemeral", false],
  ])("%s access", (_, isPersistent) => {
    beforeEach(() => {
      jest.clearAllMocks();
      mockFetchCommand.mockResolvedValue({
        ok: true,
        message: "a message",
        id: "abcefg",
        isPreexisting: false,
        isPersistent,
        event: {
          permission: {
            type: "session",
            spec: {
              resource: {
                arn: "arn:aws:ec2:us-west-2:391052057035:instance/i-0b1b7b7b7b7b7b7b7",
              },
            },
          },
        },
      });
    });

    it("should call p0 request with reason arg", async () => {
      void sshCommand(yargs())
        .fail(noop)
        .parse(`ssh some-instance --reason reason --provider aws`);
      await sleep(100);
      const hiddenFilenameRequestArgs = omit(
        mockFetchCommand.mock.calls[0][1],
        "$0"
      );
      expect(hiddenFilenameRequestArgs).toMatchSnapshot("args");
    });

    it("should wait for access grant", async () => {
      const promise = sshCommand(yargs()).fail(noop).parse(`ssh some-instance`);
      const wait = sleep(100);
      await Promise.race([wait, promise]);
      await expect(wait).resolves.toBeUndefined();
    });

    it("should wait for provisioning", async () => {
      const promise = sshCommand(yargs()).fail(noop).parse(`ssh some-instance`);
      await sleep(100); // Need to wait for listen before trigger in tests
      (onSnapshot as any).trigger({
        status: "APPROVED",
      });
      const wait = sleep(100);
      await Promise.race([wait, promise]);
      await expect(wait).resolves.toBeUndefined();
    });

    it("should call sshOrScp with non-interactive command", async () => {
      const promise = sshCommand(yargs())
        .fail(noop)
        .parse(`ssh some-instance do something`);
      await sleep(100); // Need to wait for listen before trigger in tests
      (onSnapshot as any).trigger({
        status: "APPROVED",
      });
      await sleep(100); // Need to wait for listen before trigger in tests
      (onSnapshot as any).trigger(MOCK_REQUEST);
      await expect(promise).resolves.toBeDefined();
      expect(mockPrint2.mock.calls).toMatchSnapshot("stderr");
      expect(mockPrint1).not.toHaveBeenCalled();
      expect(mockSshOrScp).toHaveBeenCalled();
    });

    it("should call sshOrScp with interactive session", async () => {
      const promise = sshCommand(yargs()).fail(noop).parse(`ssh some-instance`);
      await sleep(100); // Need to wait for listen before trigger in tests
      (onSnapshot as any).trigger({
        status: "APPROVED",
      });
      await sleep(100); // Need to wait for listen before trigger in tests
      (onSnapshot as any).trigger(MOCK_REQUEST);
      await expect(promise).resolves.toBeDefined();
      expect(mockPrint2.mock.calls).toMatchSnapshot("stderr");
      expect(mockPrint1).not.toHaveBeenCalled();
      expect(mockSshOrScp).toHaveBeenCalled();
    });
  });
});
