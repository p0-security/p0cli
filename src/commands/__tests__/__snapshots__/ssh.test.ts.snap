// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssh ephemeral access should call ssm: stderr 1`] = `
[
  [
    "a message",
  ],
  [
    "Will wait up to 5 minutes for this request to complete...",
  ],
  [
    "Your request was approved",
  ],
  [
    "Waiting for access to be provisioned",
  ],
]
`;

exports[`ssh persistent access should call ssm: stderr 1`] = `
[
  [
    "Waiting for access to be provisioned",
  ],
]
`;
